modules/consensus/tendermint/prover/src/tests/integration_tests.rs

```diff
@@ -0,0 +1,267 @@
+#[cfg(test)]
+mod tests {
+	use std::time::{SystemTime, UNIX_EPOCH};
+
+	use crate::{prove_header_update, Client, CometBFTClient, HeimdallClient};
+	use tendermint_primitives::{TrustedState, VerificationError, VerificationOptions};
+	use tracing::trace;
+
+	fn get_standard_rpc_url() -> String {
+		std::env::var("STANDARD_TENDERMINT_URL")
+			.expect("STANDARD_TENDERMINT_URL environment variable must be set")
+	}
+
+	fn get_polygon_rpc_url() -> String {
+		std::env::var("POLYGON_HEIMDALL")
+			.expect("POLYGON_HEIMDALL environment variable must be set")
+	}
+	const VALIDATOR_SET_TRANSITIONS: u32 = 4;
+
+	#[tokio::test]
+	#[ignore]
+	async fn test_standard_tendermint_integration() {
+		trace!(
+			"Testing Standard Tendermint with {} validator set transitions",
+			VALIDATOR_SET_TRANSITIONS
+		);
+		run_integration_test_standard(&get_standard_rpc_url()).await.unwrap();
+	}
+
+	#[tokio::test]
+	#[ignore]
+	async fn test_polygon_heimdall_basic_rpc() {
+		trace!("Testing Polygon's Heimdall Fork (Basic RPC)");
+		test_polygon_basic_rpc(&get_polygon_rpc_url()).await.unwrap();
+	}
+
+	// Fails with Invalid Proof, debugging
+	#[tokio::test]
+	#[ignore]
+	async fn test_polygon_heimdall_full_verification() {
+		trace!(
+			"Testing Polygon's Heimdall Fork (Full Verification) with {} validator set transitions",
+			VALIDATOR_SET_TRANSITIONS
+		);
+		run_integration_test_heimdall(&get_polygon_rpc_url()).await.unwrap();
+	}
+
+	/// Full integration test: prover and verifier for standard CometBFT with multiple validator set
+	/// transitions
+	async fn run_integration_test_standard(
+		rpc_url: &str,
+	) -> Result<(), Box<dyn std::error::Error>> {
+		let client = CometBFTClient::new(rpc_url).await?;
+		ensure_healthy(&client).await?;
+		let chain_id = client.chain_id().await?;
+		let latest_height = client.latest_height().await?;
+
+		let trusted_height = latest_height.saturating_sub(50);
+		let trusted_header = client.signed_header(trusted_height).await?;
+		let trusted_validators = client.validators(trusted_height).await?;
+		let trusted_next_validators = client.next_validators(trusted_height).await?;
+
+		let mut trusted_state = TrustedState::new(
+			chain_id,
+			trusted_height,
+			trusted_header.header.time.unix_timestamp() as u64,
+			trusted_header.header.hash().as_bytes().try_into().unwrap(),
+			trusted_validators,
+			trusted_next_validators,
+			trusted_header.header.next_validators_hash.as_bytes().try_into().unwrap(),
+			7200, // 2 hour trusting period
+			VerificationOptions::default(),
+		);
+		trusted_state.validate()?;
+
+		trace!(
+			"Starting verification with {} validator set transitions",
+			VALIDATOR_SET_TRANSITIONS
+		);
+		trace!("Initial trusted height: {}", trusted_state.height);
+
+		// Perform multiple validator set transitions
+		for transition in 1..=VALIDATOR_SET_TRANSITIONS {
```

Comment:
To test validator set transitions only count a transition when trusted_state.next_validator_hash is equal to consensus_proof.signed_header.validator_hash